src/modules/pulumi.rs

```diff
@@ -0,0 +1,295 @@
+
+    let project_file = find_package_file(&context.logical_dir)?;
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|variable, _| match variable {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "version" => {
+                    let stdout = context.exec_cmd("pulumi", &["version"])?.stdout;
+                    VersionFormatter::format_module_version(
+                        module.get_name(),
+                        parse_version(&stdout),
+                        config.version_format,
+                    )
+                }
+                .map(Ok),
+                "stack" => stack_name(&project_file, context).map(Ok),
+                _ => None,
+            })
+            .parse(None)
+    });
+
+    match parsed {
+        Ok(x) => {
+            module.set_segments(x);
+            Some(module)
+        }
+        Err(e) => {
+            log::warn!("Error in module `pulumi`:\n{}", e);
+            None
+        }
+    }
+}
+
+/// Parse the output of `pulumi version` into just the version string.
+///
+/// Normally, this just means returning it. When Pulumi is being developed, it
+/// can return results like `3.12.0-alpha.1630554544+f89e9a29.dirty`, which we
+/// don't want to see. Instead we display that as `3.12.0-alpha`.
+fn parse_version(version: &str) -> &str {
+    &version[0..version
+        .trim()
+        .split('.')
+        .take(3)
+        .map(&str::len)
+        .sum::<usize>()
+        + 2]
+}
+
+/// Find a file describing a Pulumi package in the current directory (or any parrent directory).
+fn find_package_file(path: &Path) -> Option<PathBuf> {
+    for path in path.ancestors() {
+        log::trace!("Looking for package file in {:?}", path);
+        let dir = std::fs::read_dir(path).ok()?;
+        let goal = dir.filter_map(Result::ok).find(|path| {
+            path.file_name() == OsStr::new("Pulumi.yaml")
+                || path.file_name() == OsStr::new("Pulumi.yml")
+        });
+        if let Some(goal) = goal {
+            return Some(goal.path());
+        }
+    }
+    log::trace!("Did not find a Pulumi package file");
+    None
+}
+
+/// We get the name of the current stack.
+///
+/// Pulumi has no CLI option that is fast enough to get this for us, but finding
+/// the location is simple. We get it ourselves.
+fn stack_name(project_file: &Path, context: &Context) -> Option<String> {
+    let mut file = File::open(&project_file).ok()?;
+
+    let mut contents = String::new();
+    file.read_to_string(&mut contents).ok()?;
+    let name =
+        YamlLoader::load_from_str(&contents)
+            .ok()
+            .map(|yaml| -> Option<Option<String>> {
```

Comment:
```suggestion
    .and_then(|yaml| -> Option<String> {
            log::trace!("Parsed {:?} into yaml", project_file);
            let yaml = yaml.into_iter().next()?;
            yaml.into_hash()
                .map(|mut hash| -> Option<String> {
                    hash.remove(&Yaml::String("name".to_string()))?
                        .into_string()
                })
                .flatten()
        })?;
```

I believe this section can be rewritten like this to avoid some of the issues with triple-nesting options.

(Not convinced it's significantly better, but I do find it amusing that there's a `???` at the end of this block).

Reply:
This is more of a cosmetic change, feel free to reject it!